File: church/church/report/church_attendance_report/church_attendance_report.js
// Copyright (c) 2026, kunle and contributors
// For license information, please see license.txt
/* eslint-disable */


// Attendance Analysis Report - Client Script
// Adds custom buttons for HTML preview, Excel export, PDF, and email
// Supports single sheet and date range analysis

frappe.query_reports["Church Attendance Report"] = {
    "filters": [
        {
            "fieldname": "mode",
            "label": __("Report Mode"),
            "fieldtype": "Select",
            "options": ["Single Sheet", "Date Range", "Branch Analysis"],
            "default": "Single Sheet",
            "reqd": 1,
            "on_change": function() {
                // Refresh to show/hide filters based on mode
                frappe.query_report.refresh();
            }
        },
        {
            "fieldname": "attendance_sheet",
            "label": __("Attendance Sheet"),
            "fieldtype": "Link",
            "options": "Attendance Sheet",
            "depends_on": "eval:doc.mode=='Single Sheet'",
            "mandatory_depends_on": "eval:doc.mode=='Single Sheet'",
            "get_query": function() {
                return {
                    "filters": {
                        "docstatus": ["in", [0, 1]]
                    }
                };
            }
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "depends_on": "eval:doc.mode=='Date Range'",
            "mandatory_depends_on": "eval:doc.mode=='Date Range'",
            "default": frappe.datetime.month_start()
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "depends_on": "eval:doc.mode=='Date Range'",
            "mandatory_depends_on": "eval:doc.mode=='Date Range'",
            "default": frappe.datetime.month_end()
        },
        {
            "fieldname": "branch",
            "label": __("Branch"),
            "fieldtype": "Link",
            "options": "Branch",
            "depends_on": "eval:doc.mode=='Date Range' || doc.mode=='Branch Analysis'",
            "mandatory_depends_on": "eval:doc.mode=='Branch Analysis'"
        },
        {
            "fieldname": "limit",
            "label": __("Limit"),
            "fieldtype": "Int",
            "default": 20,
            "depends_on": "eval:doc.mode=='Branch Analysis'",
            "description": __("Number of recent sheets to analyze")
        }
    ],
    
    "onload": function(report) {
        // Add custom buttons
        add_custom_report_buttons(report);
    },
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        // Highlight visitor percentage
        if (column.fieldname == "visitor_percent" && data && data.visitor_percent > 30) {
            value = `<span style="color: red; font-weight: bold;">${value}</span>`;
        }
        
        // Highlight total column
        if (column.fieldname == "total" && data && data.total > 0) {
            value = `<span style="color: #667eea; font-weight: bold;">${value}</span>`;
        }
        
        // Make sheet clickable
        if (column.fieldname == "attendance_sheet" && data && data.attendance_sheet) {
            value = `<a href="/app/attendance-sheet/${data.attendance_sheet}" target="_blank">${value}</a>`;
        }
        
        return value;
    }
};

function add_custom_report_buttons(report) {
    // Export Period to Excel Button
    report.page.add_inner_button(__('Export Period to Excel'), function() {
        const filters = report.get_values();
        export_period_to_excel(filters);
    }, __('Actions'));
    
    // For Single Sheet Mode - Add all buttons
    report.page.add_inner_button(__('Preview HTML Report'), function() {
        const filters = report.get_values();
        
        if (filters.mode === 'Single Sheet' && filters.attendance_sheet) {
            preview_html_report(filters.attendance_sheet);
        } else {
            frappe.msgprint(__('HTML preview only available in Single Sheet mode'));
        }
    }, __('Actions'));
    
    report.page.add_inner_button(__('Export to Beautiful Excel'), function() {
        const filters = report.get_values();
        
        if (filters.mode === 'Single Sheet' && filters.attendance_sheet) {
            export_to_excel(filters.attendance_sheet);
        } else {
            frappe.msgprint(__('Beautiful Excel only available in Single Sheet mode. Use "Export Period to Excel" for date ranges.'));
        }
    }, __('Actions'));
    
    report.page.add_inner_button(__('Download PDF'), function() {
        const filters = report.get_values();
        
        if (filters.mode === 'Single Sheet' && filters.attendance_sheet) {
            download_pdf_report(filters.attendance_sheet);
        } else {
            frappe.msgprint(__('PDF download only available in Single Sheet mode'));
        }
    }, __('Actions'));
    
    report.page.add_inner_button(__('Email Report'), function() {
        const filters = report.get_values();
        
        if (filters.mode === 'Single Sheet' && filters.attendance_sheet) {
            show_email_dialog(filters.attendance_sheet);
        } else {
            frappe.msgprint(__('Email feature only available in Single Sheet mode'));
        }
    }, __('Actions'));
    
    // View Attendance Sheet Button
    report.page.add_inner_button(__('View Attendance Sheet'), function() {
        const filters = report.get_values();
        
        if (filters.mode === 'Single Sheet' && filters.attendance_sheet) {
            frappe.set_route('Form', 'Attendance Sheet', filters.attendance_sheet);
        } else {
            frappe.msgprint(__('Select a single Attendance Sheet to view'));
        }
    }, __('Navigate'));
}

function export_period_to_excel(filters) {
    frappe.show_alert({
        message: __('Generating Excel file for selected period...'),
        indicator: 'blue'
    });
    
    frappe.call({
        method: 'church.church.report.attendance_analysis.attendance_analysis.export_period_to_excel',
        args: {
            filters: filters
        },
        callback: function(r) {
            if (r.message && r.message.success) {
                // Create download link
                const link = document.createElement('a');
                link.href = 'data:application/vnd.openxmlformats-officedocument.spreadsheetml.sheet;base64,' + r.message.file_content;
                link.download = r.message.filename;
                document.body.appendChild(link);
                link.click();
                document.body.removeChild(link);
                
                frappe.show_alert({
                    message: __('Excel file downloaded successfully'),
                    indicator: 'green'
                }, 5);
            } else {
                frappe.msgprint({
                    title: __('Export Failed'),
                    message: r.message ? r.message.error : __('Failed to generate Excel file'),
                    indicator: 'red'
                });
            }
        }
    });
}

function preview_html_report(attendance_sheet) {
    frappe.show_alert({
        message: __('Generating HTML report...'),
        indicator: 'blue'
    });
    
    frappe.call({
        method: 'church.church.report.attendance_analysis.attendance_analysis.generate_html_report',
        args: {
            attendance_sheet: attendance_sheet
        },
        callback: function(r) {
            if (r.message) {
                // Open in new window
                const report_window = window.open('', '_blank');
                report_window.document.write(r.message);
                report_window.document.close();
            }
        }
    });
}

function export_to_excel(attendance_sheet) {
    frappe.show_alert({
        message: __('Generating Excel file...'),
        indicator: 'blue'
    });
    
    frappe.call({
        method: 'church.church.report.attendance_analysis.attendance_analysis.export_to_beautiful_excel',
        args: {
            attendance_sheet: attendance_sheet
        },
        callback: function(r) {
            if (r.message && r.message.success) {
                // Create download link
                const link = document.createElement('a');
                link.href = 'data:application/vnd.openxmlformats-officedocument.spreadsheetml.sheet;base64,' + r.message.file_content;
                link.download = r.message.filename;
                document.body.appendChild(link);
                link.click();
                document.body.removeChild(link);
                
                frappe.show_alert({
                    message: __('Excel file downloaded successfully'),
                    indicator: 'green'
                }, 5);
            } else {
                frappe.msgprint({
                    title: __('Export Failed'),
                    message: r.message.error || __('Failed to generate Excel file'),
                    indicator: 'red'
                });
            }
        }
    });
}

function download_pdf_report(attendance_sheet) {
    frappe.call({
        method: 'church.church.report.attendance_analysis.attendance_analysis.generate_html_report',
        args: {
            attendance_sheet: attendance_sheet
        },
        callback: function(r) {
            if (r.message) {
                // Open in new window and trigger print
                const report_window = window.open('', '_blank');
                report_window.document.write(r.message);
                report_window.document.close();
                
                // Wait for content to load then print
                setTimeout(function() {
                    report_window.print();
                }, 500);
            }
        }
    });
}

function show_email_dialog(attendance_sheet) {
    // Get default recipients
    frappe.call({
        method: 'frappe.client.get_list',
        args: {
            doctype: 'Member',
            filters: { 'is_a_pastor': 1 },
            fields: ['email', 'full_name']
        },
        callback: function(r) {
            const default_recipients = r.message || [];
            const email_addresses = default_recipients
                .filter(m => m.email)
                .map(m => m.email)
                .join(', ');
            
            // Create email dialog
            const d = new frappe.ui.Dialog({
                title: __('Email Attendance Report'),
                fields: [
                    {
                        label: __('Recipients'),
                        fieldname: 'recipients',
                        fieldtype: 'Small Text',
                        reqd: 1,
                        default: email_addresses,
                        description: __('Enter email addresses separated by commas')
                    },
                    {
                        fieldtype: 'Section Break'
                    },
                    {
                        fieldtype: 'HTML',
                        fieldname: 'preview',
                        options: `
                            <div style="padding: 15px; background: #f8f9fa; border-radius: 8px; margin-top: 10px;">
                                <h4 style="margin-bottom: 10px; color: #667eea;">📧 Email Preview</h4>
                                <p><strong>Includes:</strong></p>
                                <ul style="margin-left: 20px;">
                                    <li>Summary statistics with visual cards</li>
                                    <li>Detailed breakdown by programme</li>
                                    <li>Visual analytics with progress bars</li>
                                    <li>Beautiful, professional HTML design</li>
                                </ul>
                            </div>
                        `
                    }
                ],
                primary_action_label: __('Send Email'),
                primary_action(values) {
                    const recipients_list = values.recipients
                        .split(',')
                        .map(email => email.trim())
                        .filter(email => email);
                    
                    if (recipients_list.length === 0) {
                        frappe.msgprint(__('Please enter at least one email address'));
                        return;
                    }
                    
                    d.hide();
                    
                    frappe.call({
                        method: 'church.church.report.attendance_analysis.attendance_analysis.email_report',
                        args: {
                            attendance_sheet: attendance_sheet,
                            recipients: recipients_list
                        },
                        freeze: true,
                        freeze_message: __('Sending email...'),
                        callback: function(r) {
                            if (r.message && r.message.success) {
                                frappe.show_alert({
                                    message: __('Email sent successfully'),
                                    indicator: 'green'
                                }, 5);
                            }
                        }
                    });
                }
            });
            
            d.show();
        }
    });
}

function add_custom_report_buttons(report) {
    // Preview HTML Report Button
    report.page.add_inner_button(__('Preview HTML Report'), function() {
        const filters = report.get_values();
        
        if (!filters.attendance_sheet) {
            frappe.msgprint(__('Please select an Attendance Sheet'));
            return;
        }
        
        preview_html_report(filters.attendance_sheet);
    }, __('Actions'));
    
    // Export to Beautiful Excel Button
    report.page.add_inner_button(__('Export to Beautiful Excel'), function() {
        const filters = report.get_values();
        
        if (!filters.attendance_sheet) {
            frappe.msgprint(__('Please select an Attendance Sheet'));
            return;
        }
        
        export_to_excel(filters.attendance_sheet);
    }, __('Actions'));
    
    // Download PDF Button
    report.page.add_inner_button(__('Download PDF'), function() {
        const filters = report.get_values();
        
        if (!filters.attendance_sheet) {
            frappe.msgprint(__('Please select an Attendance Sheet'));
            return;
        }
        
        download_pdf_report(filters.attendance_sheet);
    }, __('Actions'));
    
    // Email Report Button
    report.page.add_inner_button(__('Email Report'), function() {
        const filters = report.get_values();
        
        if (!filters.attendance_sheet) {
            frappe.msgprint(__('Please select an Attendance Sheet'));
            return;
        }
        
        show_email_dialog(filters.attendance_sheet);
    }, __('Actions'));
    
    // View Attendance Sheet Button
    report.page.add_inner_button(__('View Attendance Sheet'), function() {
        const filters = report.get_values();
        
        if (!filters.attendance_sheet) {
            frappe.msgprint(__('Please select an Attendance Sheet'));
            return;
        }
        
        frappe.set_route('Form', 'Attendance Sheet', filters.attendance_sheet);
    }, __('Navigate'));
}

function preview_html_report(attendance_sheet) {
    frappe.show_alert({
        message: __('Generating HTML report...'),
        indicator: 'blue'
    });
    
    frappe.call({
        method: 'church.church.report.attendance_analysis.attendance_analysis.generate_html_report',
        args: {
            attendance_sheet: attendance_sheet
        },
        callback: function(r) {
            if (r.message) {
                // Open in new window
                const report_window = window.open('', '_blank');
                report_window.document.write(r.message);
                report_window.document.close();
            }
        }
    });
}

function export_to_excel(attendance_sheet) {
    frappe.show_alert({
        message: __('Generating Excel file...'),
        indicator: 'blue'
    });
    
    frappe.call({
        method: 'church.church.report.attendance_analysis.attendance_analysis.export_to_beautiful_excel',
        args: {
            attendance_sheet: attendance_sheet
        },
        callback: function(r) {
            if (r.message && r.message.success) {
                // Create download link
                const link = document.createElement('a');
                link.href = 'data:application/vnd.openxmlformats-officedocument.spreadsheetml.sheet;base64,' + r.message.file_content;
                link.download = r.message.filename;
                document.body.appendChild(link);
                link.click();
                document.body.removeChild(link);
                
                frappe.show_alert({
                    message: __('Excel file downloaded successfully'),
                    indicator: 'green'
                }, 5);
            } else {
                frappe.msgprint({
                    title: __('Export Failed'),
                    message: r.message.error || __('Failed to generate Excel file'),
                    indicator: 'red'
                });
            }
        }
    });
}

function download_pdf_report(attendance_sheet) {
    frappe.call({
        method: 'church.church.report.attendance_analysis.attendance_analysis.generate_html_report',
        args: {
            attendance_sheet: attendance_sheet
        },
        callback: function(r) {
            if (r.message) {
                // Open in new window and trigger print
                const report_window = window.open('', '_blank');
                report_window.document.write(r.message);
                report_window.document.close();
                
                // Wait for content to load then print
                setTimeout(function() {
                    report_window.print();
                }, 500);
            }
        }
    });
}

function show_email_dialog(attendance_sheet) {
    // Get default recipients
    frappe.call({
        method: 'frappe.client.get_list',
        args: {
            doctype: 'Member',
            filters: { 'is_a_pastor': 1 },
            fields: ['email', 'full_name']
        },
        callback: function(r) {
            const default_recipients = r.message || [];
            const email_addresses = default_recipients
                .filter(m => m.email)
                .map(m => m.email)
                .join(', ');
            
            // Create email dialog
            const d = new frappe.ui.Dialog({
                title: __('Email Attendance Report'),
                fields: [
                    {
                        label: __('Recipients'),
                        fieldname: 'recipients',
                        fieldtype: 'Small Text',
                        reqd: 1,
                        default: email_addresses,
                        description: __('Enter email addresses separated by commas')
                    },
                    {
                        fieldtype: 'Section Break'
                    },
                    {
                        fieldtype: 'HTML',
                        fieldname: 'preview',
                        options: `
                            <div style="padding: 15px; background: #f8f9fa; border-radius: 8px; margin-top: 10px;">
                                <h4 style="margin-bottom: 10px; color: #667eea;">📧 Email Preview</h4>
                                <p><strong>Includes:</strong></p>
                                <ul style="margin-left: 20px;">
                                    <li>Summary statistics with visual cards</li>
                                    <li>Detailed breakdown by programme</li>
                                    <li>Visual analytics with progress bars</li>
                                    <li>Beautiful, professional HTML design</li>
                                </ul>
                            </div>
                        `
                    }
                ],
                primary_action_label: __('Send Email'),
                primary_action(values) {
                    const recipients_list = values.recipients
                        .split(',')
                        .map(email => email.trim())
                        .filter(email => email);
                    
                    if (recipients_list.length === 0) {
                        frappe.msgprint(__('Please enter at least one email address'));
                        return;
                    }
                    
                    d.hide();
                    
                    frappe.call({
                        method: 'church.church.report.attendance_analysis.attendance_analysis.email_report',
                        args: {
                            attendance_sheet: attendance_sheet,
                            recipients: recipients_list
                        },
                        freeze: true,
                        freeze_message: __('Sending email...'),
                        callback: function(r) {
                            if (r.message && r.message.success) {
                                frappe.show_alert({
                                    message: __('Email sent successfully'),
                                    indicator: 'green'
                                }, 5);
                            }
                        }
                    });
                }
            });
            
            d.show();
        }
    });
}

File: church/church/report/church_attendance_report/church_attendance_report.py
# Copyright (c) 2026, kunle and contributors
# For license information, please see license.txt

# -*- coding: utf-8 -*-
# Copyright (c) 2026, Church Management and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import getdate, formatdate, format_date, format_datetime, now_datetime, flt, add_days
import json


def execute(filters=None):
    """
    Main execution function for Script Report
    Returns columns and data
    """
    if not filters:
        filters = {}
    
    # Validate filters
    validate_filters(filters)
    
    # Get data
    columns = get_columns()
    data = get_data(filters)
    
    # Add charts
    chart = get_chart_data(filters, data)
    
    # Add summary
    report_summary = get_report_summary(filters, data)
    
    return columns, data, None, chart, report_summary


def validate_filters(filters):
    """Validate filter combinations"""
    
    # Check if at least one filter is provided
    has_sheet = filters.get("attendance_sheet")
    has_date_range = filters.get("from_date") and filters.get("to_date")
    has_branch = filters.get("branch")
    
    if not has_sheet and not has_date_range and not has_branch:
        frappe.msgprint(_("Please select either an Attendance Sheet, Date Range, or Branch"))
        return False
    
    # Validate date range
    if filters.get("from_date") and filters.get("to_date"):
        if getdate(filters.get("from_date")) > getdate(filters.get("to_date")):
            frappe.throw(_("From Date cannot be greater than To Date"))
    
    return True


def get_columns():
    """Define report columns"""
    return [
        {
            "fieldname": "attendance_sheet",
            "label": _("Sheet"),
            "fieldtype": "Link",
            "options": "Attendance Sheet",
            "width": 150
        },
        {
            "fieldname": "branch",
            "label": _("Branch"),
            "fieldtype": "Link",
            "options": "Branch",
            "width": 120
        },
        {
            "fieldname": "date",
            "label": _("Date"),
            "fieldtype": "Date",
            "width": 100
        },
        {
            "fieldname": "day",
            "label": _("Day"),
            "fieldtype": "Data",
            "width": 80
        },
        {
            "fieldname": "programme",
            "label": _("Programme"),
            "fieldtype": "Data",
            "width": 180
        },
        {
            "fieldname": "men",
            "label": _("Men"),
            "fieldtype": "Int",
            "width": 70
        },
        {
            "fieldname": "women",
            "label": _("Women"),
            "fieldtype": "Int",
            "width": 80
        },
        {
            "fieldname": "children",
            "label": _("Children"),
            "fieldtype": "Int",
            "width": 80
        },
        {
            "fieldname": "total",
            "label": _("Total"),
            "fieldtype": "Int",
            "width": 80
        },
        {
            "fieldname": "new_men",
            "label": _("New Men"),
            "fieldtype": "Int",
            "width": 90
        },
        {
            "fieldname": "new_women",
            "label": _("New Women"),
            "fieldtype": "Int",
            "width": 100
        },
        {
            "fieldname": "new_children",
            "label": _("New Children"),
            "fieldtype": "Int",
            "width": 110
        },
        {
            "fieldname": "visitors",
            "label": _("Visitors"),
            "fieldtype": "Int",
            "width": 90
        },
        {
            "fieldname": "visitor_percent",
            "label": _("Visitor %"),
            "fieldtype": "Percent",
            "width": 90
        }
    ]


def get_data(filters):
    """Get report data based on filters"""
    
    data = []
    
    # Scenario 1: Single Attendance Sheet
    if filters.get("attendance_sheet"):
        data = get_single_sheet_data(filters)
    
    # Scenario 2: Date Range (with optional branch filter)
    elif filters.get("from_date") and filters.get("to_date"):
        data = get_date_range_data(filters)
    
    # Scenario 3: Branch only
    elif filters.get("branch"):
        data = get_branch_data(filters)
    
    return data


def get_single_sheet_data(filters):
    """Get data for a single attendance sheet"""
    
    attendance_sheet = filters.get("attendance_sheet")
    doc = frappe.get_doc("Attendance Sheet", attendance_sheet)
    
    data = []
    
    for row in doc.church_attendance_analysis:
        visitor_percent = calculate_percentage(row.new_total or 0, row.total or 0)
        
        data.append({
            "attendance_sheet": doc.name,
            "branch": doc.branch,
            "date": row.date,
            "day": row.day,
            "programme": row.programme,
            "men": row.men or 0,
            "women": row.women or 0,
            "children": row.children or 0,
            "total": row.total or 0,
            "new_men": row.new_men or 0,
            "new_women": row.new_women or 0,
            "new_children": row.new_children or 0,
            "visitors": row.new_total or 0,
            "visitor_percent": visitor_percent
        })
    
    return data


def get_date_range_data(filters):
    """Get data for a date range across multiple sheets"""
    
    from_date = getdate(filters.get("from_date"))
    to_date = getdate(filters.get("to_date"))
    
    # Build query filters
    sheet_filters = {
        "reporting_date": ["between", [from_date, to_date]]
    }
    
    if filters.get("branch"):
        sheet_filters["branch"] = filters.get("branch")
    
    # Get all attendance sheets in the date range
    sheets = frappe.get_all(
        "Attendance Sheet",
        filters=sheet_filters,
        fields=["name", "branch", "reporting_date"],
        order_by="reporting_date asc"
    )
    
    data = []
    
    for sheet in sheets:
        doc = frappe.get_doc("Attendance Sheet", sheet.name)
        
        for row in doc.church_attendance_analysis:
            visitor_percent = calculate_percentage(row.new_total or 0, row.total or 0)
            
            data.append({
                "attendance_sheet": doc.name,
                "branch": doc.branch,
                "date": row.date,
                "day": row.day,
                "programme": row.programme,
                "men": row.men or 0,
                "women": row.women or 0,
                "children": row.children or 0,
                "total": row.total or 0,
                "new_men": row.new_men or 0,
                "new_women": row.new_women or 0,
                "new_children": row.new_children or 0,
                "visitors": row.new_total or 0,
                "visitor_percent": visitor_percent
            })
    
    return data


def get_branch_data(filters):
    """Get latest attendance data for a specific branch"""
    
    branch = filters.get("branch")
    limit = filters.get("limit") or 10
    
    sheets = frappe.get_all(
        "Attendance Sheet",
        filters={"branch": branch},
        fields=["name", "branch", "reporting_date"],
        order_by="reporting_date desc",
        limit=limit
    )
    
    data = []
    
    for sheet in sheets:
        doc = frappe.get_doc("Attendance Sheet", sheet.name)
        
        for row in doc.church_attendance_analysis:
            visitor_percent = calculate_percentage(row.new_total or 0, row.total or 0)
            
            data.append({
                "attendance_sheet": doc.name,
                "branch": doc.branch,
                "date": row.date,
                "day": row.day,
                "programme": row.programme,
                "men": row.men or 0,
                "women": row.women or 0,
                "children": row.children or 0,
                "total": row.total or 0,
                "new_men": row.new_men or 0,
                "new_women": row.new_women or 0,
                "new_children": row.new_children or 0,
                "visitors": row.new_total or 0,
                "visitor_percent": visitor_percent
            })
    
    return data


def get_chart_data(filters, data):
    """Generate chart data for visualization"""
    
    if not data:
        return None
    
    # Aggregate totals
    total_men = sum(row.get("men", 0) for row in data)
    total_women = sum(row.get("women", 0) for row in data)
    total_children = sum(row.get("children", 0) for row in data)
    
    # Pie chart for attendance distribution
    chart = {
        "data": {
            "labels": ["Men", "Women", "Children"],
            "datasets": [
                {
                    "name": "Attendance Distribution",
                    "values": [total_men, total_women, total_children]
                }
            ]
        },
        "type": "pie",
        "colors": ["#4facfe", "#fa709a", "#a8edea"]
    }
    
    # If date range, also show trend
    if filters.get("from_date") and filters.get("to_date"):
        chart = get_trend_chart(data)
    
    return chart


def get_trend_chart(data):
    """Generate trend chart for date range"""
    
    # Group by date
    date_totals = {}
    for row in data:
        date_key = row.get("date")
        if date_key not in date_totals:
            date_totals[date_key] = {"total": 0, "visitors": 0}
        
        date_totals[date_key]["total"] += row.get("total", 0)
        date_totals[date_key]["visitors"] += row.get("visitors", 0)
    
    # Sort by date
    sorted_dates = sorted(date_totals.keys())
    
    chart = {
        "data": {
            "labels": [format_date(d, "dd MMM") for d in sorted_dates],
            "datasets": [
                {
                    "name": "Total Attendance",
                    "values": [date_totals[d]["total"] for d in sorted_dates]
                },
                {
                    "name": "Visitors",
                    "values": [date_totals[d]["visitors"] for d in sorted_dates]
                }
            ]
        },
        "type": "line",
        "colors": ["#667eea", "#f5576c"]
    }
    
    return chart


def get_report_summary(filters, data):
    """Generate report summary cards"""
    
    if not data:
        return []
    
    # Aggregate totals
    total_attendance = sum(row.get("total", 0) for row in data)
    total_men = sum(row.get("men", 0) for row in data)
    total_women = sum(row.get("women", 0) for row in data)
    total_children = sum(row.get("children", 0) for row in data)
    total_visitors = sum(row.get("visitors", 0) for row in data)
    
    # Count unique dates/sheets
    unique_dates = len(set(row.get("date") for row in data if row.get("date")))
    unique_sheets = len(set(row.get("attendance_sheet") for row in data if row.get("attendance_sheet")))
    
    summary = [
        {
            "value": total_attendance,
            "label": _("Total Attendance"),
            "datatype": "Int",
            "indicator": "blue"
        },
        {
            "value": total_men,
            "label": _("Men"),
            "datatype": "Int",
            "indicator": "green"
        },
        {
            "value": total_women,
            "label": _("Women"),
            "datatype": "Int",
            "indicator": "orange"
        },
        {
            "value": total_children,
            "label": _("Children"),
            "datatype": "Int",
            "indicator": "purple"
        },
        {
            "value": total_visitors,
            "label": _("Visitors"),
            "datatype": "Int",
            "indicator": "red"
        },
        {
            "value": calculate_percentage(total_visitors, total_attendance),
            "label": _("Visitor %"),
            "datatype": "Percent",
            "indicator": "yellow"
        }
    ]
    
    # Add period info if date range
    if filters.get("from_date") and filters.get("to_date"):
        summary.extend([
            {
                "value": unique_dates,
                "label": _("Days Covered"),
                "datatype": "Int",
                "indicator": "gray"
            },
            {
                "value": unique_sheets,
                "label": _("Sheets Analyzed"),
                "datatype": "Int",
                "indicator": "gray"
            }
        ])
    
    return summary


def calculate_percentage(value, total):
    """Calculate percentage"""
    if not total or total == 0:
        return 0
    return round((value / total) * 100, 1)


# =============================================================================
# CUSTOM EXPORT FUNCTIONS (Updated for period support)
# =============================================================================

@frappe.whitelist()
def export_period_to_excel(filters):
    """Export date range data to Excel"""
    
    if isinstance(filters, str):
        filters = json.loads(filters)
    
    # Generate data
    columns, data, _, _, _ = execute(filters)
    
    # Create Excel file
    return create_period_excel(filters, data)


def create_period_excel(filters, data):
    """Create Excel for period analysis"""
    from frappe.utils import format_date, format_datetime
    
    try:
        import openpyxl
        from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
        from openpyxl.utils import get_column_letter
    except ImportError:
        frappe.throw(_("openpyxl library required. Run: bench pip install openpyxl"))
    
    import io
    import base64
    
    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = "Period Analysis"
    
    # Header
    ws['A1'] = "ATTENDANCE ANALYSIS - PERIOD REPORT"
    ws['A1'].font = Font(size=16, bold=True, color="FFFFFF")
    ws['A1'].fill = PatternFill(start_color="667EEA", end_color="667EEA", fill_type="solid")
    ws.merge_cells('A1:N1')
    ws['A1'].alignment = Alignment(horizontal='center', vertical='center')
    ws.row_dimensions[1].height = 30
    
    # Period info
    if filters.get("from_date") and filters.get("to_date"):
        period_text = f"{format_date(filters['from_date'], 'dd MMM yyyy')} to {format_date(filters['to_date'], 'dd MMM yyyy')}"
    else:
        period_text = "All Data"
    
    ws['A2'] = period_text
    ws['A2'].font = Font(size=12, bold=True)
    ws.merge_cells('A2:N2')
    ws['A2'].alignment = Alignment(horizontal='center')
    ws.row_dimensions[2].height = 25
    
    # Column headers
    headers = ["Sheet", "Branch", "Date", "Day", "Programme", "Men", "Women", "Children", 
               "Total", "New Men", "New Women", "New Children", "Visitors", "Visitor %"]
    
    for col_idx, header in enumerate(headers, 1):
        cell = ws.cell(row=4, column=col_idx)
        cell.value = header
        cell.font = Font(bold=True, color="FFFFFF")
        cell.fill = PatternFill(start_color="4FACFE", end_color="4FACFE", fill_type="solid")
        cell.alignment = Alignment(horizontal='center')
    
    # Data rows
    for row_idx, row_data in enumerate(data, 5):
        ws.cell(row=row_idx, column=1).value = row_data.get("attendance_sheet")
        ws.cell(row=row_idx, column=2).value = row_data.get("branch")
        ws.cell(row=row_idx, column=3).value = row_data.get("date")
        ws.cell(row=row_idx, column=4).value = row_data.get("day")
        ws.cell(row=row_idx, column=5).value = row_data.get("programme")
        ws.cell(row=row_idx, column=6).value = row_data.get("men")
        ws.cell(row=row_idx, column=7).value = row_data.get("women")
        ws.cell(row=row_idx, column=8).value = row_data.get("children")
        ws.cell(row=row_idx, column=9).value = row_data.get("total")
        ws.cell(row=row_idx, column=10).value = row_data.get("new_men")
        ws.cell(row=row_idx, column=11).value = row_data.get("new_women")
        ws.cell(row=row_idx, column=12).value = row_data.get("new_children")
        ws.cell(row=row_idx, column=13).value = row_data.get("visitors")
        ws.cell(row=row_idx, column=14).value = f"{row_data.get('visitor_percent')}%"
    
    # Save to bytes
    excel_file = io.BytesIO()
    wb.save(excel_file)
    excel_file.seek(0)
    
    file_content = base64.b64encode(excel_file.read()).decode('utf-8')
    filename = f"Attendance_Period_{filters.get('from_date', 'all')}_{filters.get('to_date', 'all')}.xlsx"
    
    return {
        'success': True,
        'file_content': file_content,
        'filename': filename
    }


@frappe.whitelist()
def export_to_beautiful_excel(attendance_sheet):
    """Export to beautifully formatted Excel (single sheet)"""
    from church.church.doctype.attendance_sheet.attendance_excel_export import export_attendance_to_excel
    
    result = export_attendance_to_excel(attendance_sheet)
    return result


@frappe.whitelist()
def generate_html_report(attendance_sheet):
    """Generate beautiful HTML report (single sheet)"""
    from church.church.doctype.attendance_sheet.attendance_report import preview_report
    
    html = preview_report(attendance_sheet)
    return html


@frappe.whitelist()
def email_report(attendance_sheet, recipients):
    """Email the report (single sheet)"""
    from church.church.doctype.attendance_sheet.attendance_report import email_report as send_email
    
    result = send_email(attendance_sheet, recipients)
    return result